test/jit/test_hooks_modules.py

```diff
@@ -0,0 +1,531 @@
+        return output + "_fwh1"
+
+    def forward_hook2(self, input: Tuple[str], output: str):
+        assert self.name == "inner_mod_name"
+        assert input == ("pre_hook_override_name2",)
+        assert output == "pre_hook_override_name2_inner_mod_fwh1"
+        return output
+
+    m.submodule.register_forward_pre_hook(pre_hook1)
+    m.submodule.register_forward_pre_hook(pre_hook2)
+    m.submodule.register_forward_hook(forward_hook1)
+    m.submodule.register_forward_hook(forward_hook2)
+
+    return m
+
+
+def test_forward_tuple_input_model():
+    # Test case where forward is passed a single tuple for input.
+    # This is different because eager always wraps pre-hook return arguments
+    # in a tuple when the returned pre-hook result isn't a tuple
+    # (to allow the result to be passed to another pre-hook if needed).
+    # The eager behavior doesn't wrap the single tuple input pre-hook return in a
+    # tuple as it should. To get consitent behavior between single tuple inputs and
+    # the rest of the possible forward inputs, pre-hooks need to
+    # wrap single tuple inputs returns in another tuple. This is
+    # enforced by the schema checker.
+    m = ModuleForwardTupleInput("outer_mod_name", "inner_mod_name")
+
+    def pre_hook_outermod(self, input: Tuple[Tuple[int]]) -> Tuple[Tuple[int]]:
+        # 'return (11,)' doesn't work with eager, inner tuple lost
+        return ((11,),)
+
+    def pre_hook_innermod(self, input: Tuple[Tuple[int]]) -> Tuple[Tuple[int]]:
+        # 'return (22,)' doesn't work with eager, inner tuple lost
+        return ((22,),)
+
+    def forward_hook_outermod(self, input: Tuple[Tuple[int]], output: int):
+        return (11,)
+
+    def forward_hook_innermod(self, input: Tuple[Tuple[int]], output: Tuple[int]):
+        return 22
+
+    m.register_forward_pre_hook(pre_hook_outermod)
+    m.submodule.register_forward_pre_hook(pre_hook_innermod)
+    m.register_forward_hook(forward_hook_outermod)
+    m.submodule.register_forward_hook(forward_hook_innermod)
+
+    return m
+
+
+def test_submodule_direct_forward_invocation_model():
+    m_submod_forward_call = ModuleDirectFowardSubmodCall(
+        "outer_mod_name", "inner_mod_name"
+    )
+    m_submod_call = ModuleForwardSingleInput("outer_mod_name", "inner_mod_name")
+
+    def pre_hook(self, input: Tuple[str]) -> Tuple[str]:
+        return ("pre_hook_override_name",)
+
+    def forward_hook(self, input: Tuple[str], output: str):
+        assert input == ("pre_hook_override_name",)
+        return output + "_fh"
+
+    m_submod_forward_call.submodule.register_forward_pre_hook(pre_hook)
+    m_submod_forward_call.submodule.register_forward_hook(forward_hook)
+    m_submod_call.submodule.register_forward_pre_hook(pre_hook)
+    m_submod_call.submodule.register_forward_hook(forward_hook)
+
+    m_submod_forward_call_scripted = torch.jit.script(m_submod_forward_call)
+    m_submod_call_scripted = torch.jit.script(m_submod_call)
+
+    return m
```

Comment:
Where is `m` defined? Why are the scripted versions of `m_submod_forward_call` and  `m_submod_call` not used?

Reply:
I also noticed that this test is absent from the C++ tests stacked on top. Why is that?

Reply:
It's actually not used anywhere, I think it got made somehow in the code refactoring at some point. I'll just delete it